### PR TITLE
DB layer rework

### PR DESCRIPTION
**Pull request by tib**
- table migration updates
- naming conventions
- update sqlkit dependency
- add testing tools
- add unique constraint support
- add pk constraint
- database join implementation

### CODE REVIEW

File: Sources/FeatherDatabase/Database/DatabaseGroupFilterInterface.swift [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                return p
+            }
+        }
+        return res
+    }
+}
+
+extension SQLSelectBuilder {
+
+    func applyFilter(
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/DatabaseGroupFilterInterface.swift:61]: In practice it is an API guarantee that you do _not_ have to actually keep track of the return values of `Self`-returning methods on any builder (with the sole exception of a `.returning()` clause, which is a special case due to legacy design constraints); as such, you can significantly simplify this method.

File: Sources/FeatherDatabase/Database/DatabaseTable.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    public static func create(on db: Database) async throws {
+        let sql = db.sqlDatabase.create(table: tableName)
+        for constraint in constraints {
+            sql.createTable.tableConstraints.append(constraint.sqlConstraint)
+        }
+        try await sql.build(columns).run()
+    }
+
+    public static func drop(on db: Database) async throws {
+        try await db.sqlDatabase.drop(table: tableName).run()
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/DatabaseTable.swift]: Depending on what semantics you want to provide by default, you may want to use the `.ifExists()` and/or `.cascade()` modifiers here (both silently do nothing if the dialect doesn't support them).

File: Sources/FeatherDatabase/Database/Query/DatabaseQueryCount.swift [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        on db: Database
+    ) async throws -> UInt {
+        try await db.run { sql in
+            try await sql
+                .select()
+                .from(Row.tableName)
+                .column(SQLFunction("COUNT"), as: "count")
+                .applyFilter(filter)
+                .first(decoding: RowCount.self)?
+                .count ?? 0
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/Query/DatabaseQueryCount.swift]:
```suggestion
                .first(decodingColumn: "count", as: UInt.self) ?? 0
```

File: Sources/FeatherDatabase/Database/Query/DatabaseQueryCount.swift [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    public static func count(
+        filter: DatabaseFilter<Row.ColumnNames>? = nil,
+        on db: Database
+    ) async throws -> UInt {
+        try await db.run { sql in
+            try await sql
+                .select()
+                .from(Row.tableName)
+                .column(SQLFunction("COUNT"), as: "count")
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/Query/DatabaseQueryCount.swift:28]:
```suggestion
                .column(SQLFunction("COUNT", args: SQLLiteral.all), as: "count")
```

File: Sources/FeatherDatabase/Database/Query/DatabaseQueryInsert.swift [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    fileprivate func chunked(into size: Int) -> [[Element]] {
+        stride(
+            from: 0,
+            to: count,
+            by: size
+        )
+        .map {
+            Array(self[$0..<Swift.min($0 + size, count)])
+        }
+    }
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/Query/DatabaseQueryInsert.swift]: I recommend adding a dependency on `swift-algorithms` and using its `chunks(ofCount:)` method; it works with all Collections (not just Array) and is significantly more efficient than this implementation.

File: Sources/FeatherDatabase/Database/Query/DatabaseQueryJoin.swift [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        filter: DatabaseFilter<T.ColumnNames>? = nil,
+        on db: Database
+    ) async throws -> [T] {
+        try await db.run { sql in
+            try await sql
+                .select()
+                .column(
+                    SQLColumn(
+                        SQLLiteral.all,
+                        table: SQLIdentifier(T.tableName)
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/Query/DatabaseQueryJoin.swift:35]: You shouldn't need the explicit `SQLIdentifier()` initializer here, `T.tableName` should work.

File: Sources/FeatherDatabase/Database/Query/DatabaseQueryJoin.swift [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                    )
+                )
+                .from(Row.tableName)
+                .join(
+                    T.tableName,
+                    method: join.method,
+                    on: SQLColumn(join.column.rawValue),
+                    join.op,
+                    SQLColumn(join.otherColumn.rawValue)
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/Query/DatabaseQueryJoin.swift]: A join condition would usually make no sense without explicit table references included for the columns, is the omission here intentional?

File: Sources/FeatherDatabase/Database/Query/DatabaseQueryList.swift [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    public static func list(
+        _ query: DatabaseList<Row.ColumnNames>,
+        on db: Database
+    ) async throws -> DatabaseList<Row.ColumnNames>.Result<Row> {
+        try await db.run { sql in
+            let total =
+                try await sql
+                .select()
+                .from(Row.tableName)
+                .column(SQLFunction("COUNT"), as: "count")
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/Query/DatabaseQueryList.swift]:
```suggestion
                .column(SQLFunction("COUNT", args: SQLLiteral.all), as: "count")
```

File: Sources/FeatherDatabase/Database/Query/DatabaseQueryList.swift [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                .applyFilter(query.filter)
+                .applyOrders(query.orders)
+                .first(decoding: RowCount.self)?
+                .count ?? 0
```
Comment (gwynne) [inline Sources/FeatherDatabase/Database/Query/DatabaseQueryList.swift]:
```suggestion
                .first(decodingColumn: "count", as: UInt.self) ?? 0
```

File: Tests/FeatherDatabaseTests/_Utils/TestCase.swift [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+import NIO
+import XCTest
+
+class TestCase: XCTestCase {
+
+    var eventLoopGroup: EventLoopGroup!
+    var components: ComponentRegistry!
+
+    override func setUp() async throws {
+        self.eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
```
Comment (gwynne) [inline Tests/FeatherDatabaseTests/_Utils/TestCase.swift:18]: It's usually recommended to use `MultiThreadedEventLoopGroup.singleton` nowadays (which also doesn't have to be explicitly shut down).

File: Tests/FeatherDatabaseTests/_Utils/XCTAssertThrowsAsync.swift [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+//
+
+import XCTest
+
+func XCTAssertThrowsAsync<E: Error>(
+    _ expression: () async throws -> Void,
+    _: E.Type,
+    _ errorBlock: ((E) async throws -> Void)? = nil,
+    _ message: String,
+    file: StaticString = #file,
```
Comment (gwynne) [inline Tests/FeatherDatabaseTests/_Utils/XCTAssertThrowsAsync.swift]:
```suggestion
    file: StaticString = #filePath,
```

File: Tests/FeatherDatabaseTests/_Utils/XCTAssertThrowsAsync.swift [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+//  Created by Tibor Bodecs on 29/02/2024.
+//
+
+import XCTest
+
+func XCTAssertThrowsAsync<E: Error>(
+    _ expression: () async throws -> Void,
+    _: E.Type,
+    _ errorBlock: ((E) async throws -> Void)? = nil,
+    _ message: String,
```
Comment (gwynne) [inline Tests/FeatherDatabaseTests/_Utils/XCTAssertThrowsAsync.swift]:
```suggestion
    _ message: @autoclosure () -> String,
```
(to match XCTest's behavior)

**Review by gwynne — CHANGES_REQUESTED** [review]
See inline comments

File: Sources/FeatherDatabase/Database/DatabaseGroupFilterInterface.swift (hunk 1 above)
Comment (tib) [reply Sources/FeatherDatabase/Database/DatabaseGroupFilterInterface.swift:61]: It's good to know this, will improve the result builders in a later version.

**Review by mzpx — APPROVED** [review]

**Pull request merged**